boot/freeldr/freeldr/freeldr.c

```diff
@@ -40,6 +40,107 @@ CCHAR FrLdrBootPath[MAX_PATH] = "";
 
 /* FUNCTIONS ******************************************************************/
 
+static
+BOOLEAN
+LoadRosload(
+    _In_ PCSTR RosloadPath,
+    _Out_ PCHAR FullPath,
+    _Out_ PVOID* ImageBase)
+{
+    BOOLEAN Success;
+
+    /* Create full rosload.exe path */
+    strcpy(FullPath, FrLdrBootPath);
+    strcat(FullPath, "\\");
+    strcat(FullPath, RosloadPath);
+
+    TRACE("Loading second stage loader '%s'\n", FullPath);
+
+    /* Load rosload.exe as a bootloader image */
+    Success = PeLdrLoadImageEx(FullPath,
+                               LoaderLoadedProgram,
+                               ImageBase,
+                               FALSE);
+    if (!Success)
+    {
+        ERR("Failed to load second stage loader '%s'\n", FullPath);
+    }
+
+    return Success;
+}
+
+static
+ULONG
+LaunchSecondStageLoader(VOID)
+{
+    CHAR SecondStageLdrPath[MAX_PATH];
+    PLDR_DATA_TABLE_ENTRY FreeldrDTE, RosloadDTE;
+    LIST_ENTRY ModuleListHead;
+    PVOID ImageBase;
+    LONG (*EntryPoint)(VOID);
+    BOOLEAN Success;
+
+    /* Initialize the loaded module list */
+    InitializeListHead(&ModuleListHead);
+
+    /* Load the second stage loader */
+    if (!LoadRosload("rosload.exe", SecondStageLdrPath, &ImageBase))
+    {
+        /* Try in loader directory */
+        if (!LoadRosload("loader\\rosload.exe", SecondStageLdrPath, &ImageBase))
```

Comment:
Actually I'm thinking it could be of interest to move rosload (and test it for this 2nd test) into the systemroot\system32 directory; what do you think about that?

Reply:
Windows has winload.exe in system32, too, so I guess that makes sense.